en/developer/customization/articles/50-liferay-forms/02-form-storage-adapters.markdown

```diff
@@ -39,61 +36,166 @@ their creation, and a different storage adapter cannot be selected.
-dependent on your own application of the storage adapter). In addition, retrieve
-and delete the DDM class storage link using `DDMStorageLinkLocalService`.
-
-`doDeleteByDDMStrcuture`
-: When a delete request is made on an entire form, delete all the form records
-associated with it. In addition, take the form's `ddmStructureId` and delete all
-the DDM structure storage links that were created for it.
+![Figure 1: Choose from available Storage Types.](../../images/forms-storage-type.png)
 
+## The `DDMStorageAdapter` Interface
+
+You must implement three methods: `delete`, `get`, and `save`.
+
+```java
+public DDMStorageAdapterDeleteResponse delete(
+        DDMStorageAdapterDeleteRequest ddmStorageAdapterDeleteRequest)
+    throws StorageException;
+
+public DDMStorageAdapterGetResponse get(
+        DDMStorageAdapterGetRequest ddmStorageAdapterGetRequest)
+    throws StorageException;
+
+public DDMStorageAdapterSaveResponse save(
+        DDMStorageAdapterSaveRequest ddmStorageAdapterSaveRequest)
+    throws StorageException;
+```
+
+Each method must return a
+_DDMStorageAdapter[[Save](https://github.com/liferay/liferay-portal/blob/7.2.0-ga1/modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/storage/DDMStorageAdapterSaveResponse.java)/[Get](https://github.com/liferay/liferay-portal/blob/7.2.0-ga1/modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/storage/DDMStorageAdapterGetResponse.java)/[Delete](https://github.com/liferay/liferay-portal/blob/7.2.0-ga1/modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/storage/DDMStorageAdapterDeleteSaveResponse.java)]Response_
+object, each constructed using a static inner `Builder` class's `newBuilder`
+method. 
+
+Each method of the storage adapter is passed a very useful
+`DDMStorageAdapter[Save/Delete/Get]Request`. The request objects contain getter
+methods that return useful contextual information.
+
+You must only worry about your implementation needs and building/returning the
+appropriate response objects. The Forms application's service code looks up the
+appropriate storage adapter, and calls its code against the
+`DDMFormInstanceRecord` for the backing form is configured to use the storage
+adapter.
+
+From `DDMFormInstanceRecordLocalServiceImpl`:
+
+```java
+DDMStorageAdapter ddmStorageAdapter = getDDMStorageAdapter();
+
+DDMStorageAdapterSaveResponse ddmStorageAdapterSaveResponse =
+    ddmStorageAdapter.save(ddmStorageAdapterSaveRequest);
+```
+
+## `DDMStorageAdapter[Save/Delete/Get]Request` Getters
+
+Call the request object's getters to access the necessary data:
+
+`DDMStorageAdapterSaveRequest`:
+
+- Method: `public String getClassName()` 
+    Returns a String, `com.liferay.dynamic.data.mapping.model.DDMStorageLink`
+    <!-- is this always returned?-->
```

Comment:
@natocesarrego this String is what I saw returned, will it always be the case?
`com.liferay.dynamic.data.mapping.model.DDMStorageLink`

Reply:
Yes, at this moment, this is the value that will be returned.